Broken link 2

antonilol opened the pull request:
test gh ci

niftynei committed:
interactive-tx: Add dual-funding flow, using the interactive tx protocol

This commit adds the interactive transaction construction protcol, as
well as the first practical example of using it, v2 of channel
establishment.

Note that for v2 we also update the channel_id, which now uses the hash
of the revocation_basepoints. We move away from using the funding
transaction id, as the introduction of RBF* makes it such that a single
channel may have many funding transaction id's over the course of
its lifetime.

*Later, also splicing

niftynei committed:
v2 opens: add a `tx_abort` message

Requested-by: @t-bast

rustyrussell committed:
BOLT 4: Remove legacy format, make var_onion_optin compulsory.

My measurements a few weeks ago reveal that only 5 nodes do not
advertize this feature, of over 17000.  I have a patch to
remove support from c-lightning, too.

[ 6 months later: t-bast notes that they only see 0.2% of htlcs using
  legacy, and my node hasn't seen one for 2 months w/ 12000 htlcs --RR ]

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

t-bast committed:
Route Blinding

Route blinding allows a recipient to provide a blinded route to
potential payers. Each node_id in the route is tweaked, and dummy
hops may be included.

This is an alternative to rendezvous to preserve recipient anonymity.
It has a different set of trade-offs: onions are re-usable, but the
privacy guarantees are a bit weaker and require more work (e.g. when
handling payment fees and errors).

t-bast committed:
Bolt 4: add route blinding construction

Add specification requirements for creating and using blinded routes.
This commit contains the low-level details of the route blinding scheme,
decoupled from how it can be used by high-level components such as onion
messages or payments.

t-bast committed:
Bolt 4: add blinded payments

Add specification requirements for using route blinding to make payments
while preserving recipient anonymity. Implementers must ensure they
understand all those requirements, there are subtle attacks that could let
malicious senders deanonymize the route if incompletely implemented.

antonilol committed:
fix broken link

vincenzopalazzo commented on 03-transactions.md:
```diff
@@ -12,7 +12,7 @@ This details the exact format of on-chain transactions, which both sides need to
           * [`to_local` Output](#to_local-output)
           * [`to_remote` Output](#to_remote-output)
-          * [`to_local_anchor` and `to_remote_anchor`](#to_local_anchor-and-to_remote_anchor-output-option_anchor_outputs)
+          * [`to_local_anchor` and `to_remote_anchor`](#to_local_anchor-and-to_remote_anchor-output-option_anchors)
```
have you tried to remove this and check if the CI will fail?

antonilol replied:
e0f6d5b

antonilol replied:
it also did not catch this one

vincenzopalazzo replied:
Amazing a url checker that is not able to check the relative url! so maybe you can run your scrip inside the CI?

joostjager committed:
BOLT 04: Onion failure message may be followed by a tlv stream

t-bast committed:
Update onion test vectors

To only use valid tlv payloads instead of fixed-size legacy ones and
invalid tlv streams.

[ Minor typo change: third payload is 275 not 256 bytes long --RR ]

joostjager committed:
BOLT 04: remove associated data from test vector

Data is not relevant for failure message generation.

joostjager committed:
BOLT 04: remove max hops from test vector

With the tlv payload, the maximum isn't fixed anymore.

joostjager committed:
BOLT 04: Update recommended failure message length and test vector

t-bast committed:
Bolt 7: restore htlc_maximum_msat requirement (#1029)

Requirements for the htlc_maximum_msat field in channel_update were
inadvertently removed by #999 (this PR meant to make this field mandatory,
not removed explanations about what it does).

t-bast committed:
Add `require_confirmed_inputs` TLV

This lets any side of the protocol require the other side to use confirmed
inputs, to avoid paying the fees of a low feerate unconfirmed ancestor.

t-bast committed:
Clarify RBF funding requirements

TheBlueMatt committed:
Allow nodes to overshoot the MPP `total_msat` when paying (#1031)

When a node retires a failed path as part of a larger MPP payment,
the node may wish to use a path which is constrained by an
`htlc_minimum_msat` value. In this case, the node is forced to
overpay, likely overshooting the `total_msat` it set in the earlier
onions for the same MPP payment.

There are two possible solutions to this - either allow the
`total_msat` value to change in later HTLCs or allow the node to
(slightly) overshoot the `total_msat` value.

Allowing `total_msat` to change across HTLCs is nontrivial to
implement - HTLCs may arrive out-of-order, causing the receiving
node to have to track all seen `total_msat` values and accept a
set of HTLCs which meet any of the seen `total_msat` values.

Instead, this commit changes the MPP logic to simply allow a sender
to overshoot the stated `total_msat`.

Sadly the backwards-compatibility story for this is not great.
There doesn't seem to be a good way to resolve this issue in a
backwards-compatible way. Instead we just bite the bullet and make
the incompatible change, hoping the overshooting is rare enough
that it's not a major issue.

t-bast committed:
Allow nodes to overshoot final htlc amount and expiry (#1032)

When nodes receive HTLCs, they verify that the contents of those HTLCs
match the intructions that the sender provided in the onion. It is
important to ensure that intermediate nodes and final nodes have similar
requirements, otherwise a malicious intermediate node could easily probe
whether the next node is the final recipient or not.

Unfortunately, the requirements for intermediate nodes were more lenient
than the requirements for final nodes. Intermediate nodes allowed overpaying
and increasing the CLTV expiry, whereas final nodes required a perfect
equality between the HTLC values and the onion values.

This provided a trivial way of probing: when relaying an HTLC, nodes could
relay 1 msat more than what the onion instructed (or increase the outgoing
expiry by 1). If the next node was an intermediate node, they would accept
this HTLC, but if the next node was the recipient, they would reject it.

We update those requirements to fix this probing attack vector.

We also clarify `min_final_cltv_expiry`: this is actually a cltv_expiry_delta,
not an absolute cltv_expiry, so the field name should reflect that.

Recipients require incoming HTLC expiry to comply with that expiry delta.

t-bast committed:
Use onion amount in MPP set calculation (#1040)

* Use onion amount in MPP set calculation

The sender chooses the amounts that are set in the onion payload
(`amt_to_forward`) but cannot predict what amounts will be set in the
HTLCs (`amount_msat`) since intermediate nodes are allowed to overpay.

* Fix error requirements for final node

These requirements were missed when integrating #1032

Roasbeef committed:
Merge pull request #1021 from joostjager/failure-message-tlv

TLV failure message and recommended length to 1024

niftynei committed:
peer-protocol: send first+second commit point in openchanv2 + accept

Repeat the second commit point in the initial openchannel2 +
acceptchannel2 messages.

Suggested-By: @pm47

niftynei committed:
v2 opens: proposal to get rid of the minimum estimated fee

Prior versions of the v2 dual-funding protocol assumed a 'minimum fee'
payment for any witness stack of any input, as a way to simplify fee
checks.

The suggested min feerate didn't make sense for taproot spend paths etc;
instead we remove this check entirely.

JssDWt committed:
A note on the maximum size of Tagged Fields (#1042)

Make it more obvious to readers of bolt 11 that the maximum length of a Tagged
Field is 639 bytes.

t-bast committed:
Add section about liquidity griefing

This issue is non-trivial and worth mentioning, otherwise implementations
may forget to handle this which would result in an easy way of attacking
node's on-chain liquidity, creating a large opportunity cost.

ellemouton committed:
BOLT 03: fix no htlc single anchor test

The commitment transaction tests are all meant to use the same funding
transaction which has an amount of 10000000000 msat. The LocalBalance
and RemoteBalance along with the value of any htlcs should always add up
to this amount.

This commit updates the `simple commitment tx with no HTLCs and single
anchor` anchors test to comply with the above.

ellemouton committed:
BOLT 03: fix legacy same amt and pre-image test vector

The `commitment tx with 3 htlc outputs, 2 offered having the same amount
and preimage` test was not correctly updated after the value of test
htlc 6 was changed to 5000001 and the cltv expiry of test htlc 5 was
changed to 506. This commit updates the legacy test accordingly.

ellemouton committed:
BOLT 03: fix static-remote same amt and pre-image test vector

The `commitment tx with 3 htlc outputs, 2 offered having the same amount
and preimage` test was not correctly updated after the value of test
htlc 6 was changed to 5000001 and the cltv expiry of test htlc 5 was
changed to 506. This commit updates the static-remote test accordingly.

ellemouton committed:
BOLT 03: fix anchors same amt and pre-image test vector

The `commitment tx with 3 htlc outputs, 2 offered having the same amount
and preimage` test was not correctly updated after the value of test
htlc 6 was changed to 5000001 and the cltv expiry of test htlc 5 was
changed to 506. This commit updates the anchors test accordingly.

wpaulino committed:
Add test vectors for option_anchors_zero_fee_htlc_tx (#1018)

These are the same test vectors as those found under Appendix F, except
that each HTLC has a zero fee transaction instead, resulting in a
signature change.

t-bast committed:
Store state when sending `commitment_signed`

If we only store state when sending `tx_signatures`, there are cases where
we cannot reconcile states if a disconnection occurs during the signing
steps: one side will have sent `tx_signatures` and thus must wait for the
transaction to be spent or double-spent, while the other side has already
forgotten that channel because they haven't sent `tx_signatures`.

This is fixed by storing state when sending `commitment_signed`, and
adding a `next_funding_txid` field to `channel_reestablish` to ask our
peer to retransmit signatures that we haven't received.

antonilol committed:
fix typos

TheBlueMatt committed:
Update onion errors since we allow overpaying or under-CLTVing

In #1032 we allowed overshooting the final amount and expiry, but
forgot to update the onion error descriptions which make reference
thereto.

dunxen committed:
Use bitcoin wire encoding for witnesses

rustyrussell committed:
BOLT 4: onion message support.

These use onion encoding for simple one-way messaging: there are no error returns.
However, every onion uses route blinding *even if it doesn't need to*.

You can prove what path was used to reach you by including `path_id` in the
encrypted_data_tlv.

Note that this doesn't actually define the payload we're transporting:
that's explictly defined to be payloads in the 64-255 range.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ddustin committed:
Bolt 2: Make interactive-tx explicitly use SIGHASH_ALL

This was previously assumed but adding it to the spec makes it explicit, should we ever want to change it in the future.

Suggested-By: @morehouse

t-bast committed:
Use signed amounts in RBF messages

While dual funding only needs unsigned funding amounts, other protocols
that leverage interactive-tx may use signed funding amounts, for example
to take funds out of an existing channel (splice-out).

It is thus more future-proof to use signed amounts in `tx_init_rbf` and
`tx_ack_rbf`.

TheBlueMatt committed:
Fix undeclared reference in onion-routing

TheBlueMatt committed:
Correct final_cltv handling in blinded paths

When paying a blinded path, we don't have a CLTV delta at each hop
available, but rather only a total CLTV delta for the entire
blinded path.

However, the onion format currently still requires that we specify
an `outgoing_cltv_value` for the final hop. As the sender, we don't
have a sensible value to put there, as we don't know which part of
the total CLTV delta belongs to the recipient.

The sender is instructed to use the values that are known to them
when setting `outgoing_cltv_value` for the final hop:
- The current block height.
- Any additional delta added to account for block propagation and
  improve privacy.

This change reflects the behavior of some implementations at the time
of writing.

rustyrussell committed:
tools: remove option handling now last optional field was removed.

6fee63fc342736a2eb7f6e856ae0d85807cc50ec made `htlc_maximum_msat`
compulsory, which was the last "optional" field.  All future options
will be TLVs, so we never need to parse the "] (optionxxx)" format
for fields again.

We allow options for messages, but only for documentation (we have
some messages which are only for `gossip_queries`), so we don't
put it in the output.

Here's the before/after diff of the output for all the spec files:

```diff
--- /tmp/before	2023-04-27 14:21:07.417170213 +0930
+++ /tmp/after	2023-04-27 14:19:39.184573086 +0930
@@ -273,7 +273,7 @@
 msgdata,channel_update,fee_base_msat,u32,
 msgdata,channel_update,fee_proportional_millionths,u32,
 msgdata,channel_update,htlc_maximum_msat,u64,
-msgtype,query_short_channel_ids,261,gossip_queries
+msgtype,query_short_channel_ids,261
 msgdata,query_short_channel_ids,chain_hash,chain_hash,
 msgdata,query_short_channel_ids,len,u16,
 msgdata,query_short_channel_ids,encoded_short_ids,byte,len
@@ -281,17 +281,17 @@
 tlvtype,query_short_channel_ids_tlvs,query_flags,1
 tlvdata,query_short_channel_ids_tlvs,query_flags,encoding_type,byte,
 tlvdata,query_short_channel_ids_tlvs,query_flags,encoded_query_flags,byte,...
-msgtype,reply_short_channel_ids_end,262,gossip_queries
+msgtype,reply_short_channel_ids_end,262
 msgdata,reply_short_channel_ids_end,chain_hash,chain_hash,
 msgdata,reply_short_channel_ids_end,full_information,byte,
-msgtype,query_channel_range,263,gossip_queries
+msgtype,query_channel_range,263
 msgdata,query_channel_range,chain_hash,chain_hash,
 msgdata,query_channel_range,first_blocknum,u32,
 msgdata,query_channel_range,number_of_blocks,u32,
 msgdata,query_channel_range,tlvs,query_channel_range_tlvs,
 tlvtype,query_channel_range_tlvs,query_option,1
 tlvdata,query_channel_range_tlvs,query_option,query_option_flags,bigsize,
-msgtype,reply_channel_range,264,gossip_queries
+msgtype,reply_channel_range,264
 msgdata,reply_channel_range,chain_hash,chain_hash,
 msgdata,reply_channel_range,first_blocknum,u32,
 msgdata,reply_channel_range,number_of_blocks,u32,
@@ -310,7 +310,7 @@
 subtype,channel_update_checksums
 subtypedata,channel_update_checksums,checksum_node_id_1,u32,
 subtypedata,channel_update_checksums,checksum_node_id_2,u32,
-msgtype,gossip_timestamp_filter,265,gossip_queries
+msgtype,gossip_timestamp_filter,265
 msgdata,gossip_timestamp_filter,chain_hash,chain_hash,
 msgdata,gossip_timestamp_filter,first_timestamp,u32,
 msgdata,gossip_timestamp_filter,timestamp_range,u32,
```

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

t-bast committed:
Remove requirements to disconnect on warnings

We generally shouldn't disconnect when sending or receiving warning
messages. Whenever disconnecting after a warning makes sense, it should
be specified in the requirements linked to that specific scenario.

Fixes #1072

nepet committed:
Bolt 1: Clarify the definition of tu fields (#1073)

Truncated integer fields, e.g. tu16 cannot have leading zeros
judging by the implementations of the parsers, which are used
across the different ln implementations.
The `can` imposed that it is not necessary to truncate the fields
to values of information.

Signed-off-by: Peter Neuroth <pet.v.ne@gmail.com>

TheBlueMatt committed:
Merge pull request #1075 from t-bast/remove-disconnect-warning

Remove requirements to disconnect on warnings

ellemouton committed:
Bolt 4: clarify onion test payload contents (#1077)

Adds a comment to the `onion-test.json` file to clarify that the
payloads specified for each hop in the test already include the variable
length encodings.

rustyrussell committed:
fixup! BOLT 4: onion message support.

Typo fixes from @t-bast, @thomash-acinq and @remyers.

rustyrussell committed:
fixup! BOLT 4: onion message support.

@thomash-acinq points out:

1. We absolutely can put other fields in `encrypted_data_tlv`, esp. padding, and test vectors do this.
2. Presumably it was supposed to refer to onionmsg_tlv, so fix that.
3. And of course we need to allow payload fields!

rustyrussell committed:
BOLT 8: make it clear that there are *two* chaining keys.

lnmessage got this wrong!  It would pass our test vectors, but actually fail
in real usage, since it used the same `ck`.

Also, nonce rotation happens after 1000 encryptions, which is when the nonce
reaches 1000 (since it's zero based!), not when it *exceeds* 1000.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ariard committed:
Specify max HTLC nLocktime for expiry_too_far

tnull committed:
Fix `short_chanel_id` typos (#1087)

carlaKC committed:
bolt07: enforce htlc_maximum_msat >= htlc_minimum_msat (#1089)

vincenzopalazzo committed:
bol09: Specify behavior when a node specifies both optional and required features

While reviewing a patch on lnprototest, I encountered a scenario
where the BOLT 9 specification needed to provide clear guidance.

As a result, this commit adds specific requirements to
determine the appropriate behaviour when a node specifies
both optional and required features.

Additionally, if this situation appears to be an
implementation bug, it will be taken care of accordingly.

Reported-by: lnprototest
Signed-off-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

t-bast committed:
Allow unset `onion_hash` in `invalid_onion_blinding` (#1093)

In most cases the `onion_hash` isn't actionable in the case of blinded
payments and it's wasteful to keep track of the incoming onion, so we
allow setting it to an all zero value.

dni committed:
Bolt11: min_final_cltv_expiry_delta is optional, not required (#1100)

valentinewallace committed:
Clarify computation of final outgoing_expiry in route blinding (#1069)

BOLT 12 invoices do not include a max_cltv_expiry field, so it's good to
clarify how senders can compute the final outgoing cltv expiry.

Roasbeef committed:
Merge pull request #1095 from vincenzopalazzo/macros/specify-multiple-fields

bol09: Specify behavior when a node specifies both optional and required features

ariard committed:
Add a `max_dust_htlc_exposure_msat` (#919)

* Bound exposure to trimmed in-flight HTLCs
* Reject update_fee beyond max_dust_htlc_exposure_msat

Co-authored-by: t-bast <bastuc@hotmail.fr>

ProofOfKeags committed:
BOLT2: change "reject" to "fail" in channel opening requirements (#1104)

Since "reject" isn't defined anywhere in the spec and "fail" does
have a definition laid out in BOLT1 that includes special
provisions for channels that have not yet been established, BOLT2
is amended to clarify the requirements for the receiver of
`accept_channel`

rustyrussell committed:
CONTRIBUTING.md: modern feature bit assignment.

1. Put it in the PR title so everyone can see.
2. Deploy with +100 while it's still unratified, in case it changes.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 1: define what `offered` and `negotiated` mean.

i.e. it was present in the init feature bits.  We use this in several places, but assume everyone knows what it means.

It's particularly fraught with required features: it's explicitly legitimate to assume these are accepted if
the node keeps talking to you after init!

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ProofOfKeags committed:
Clarify the semantics of max_htlc_value_in_flight_msat

Prior wording of the description of this parameter left room for
ambiguity around whether it capped the total value offered by both
channel peers combined or if it was solely capping the total value
of HTLCs offered by the remote.

Roasbeef committed:
Merge pull request #1109 from rustyrussell/define-negotiated

BOLT 1: define what `offered` and `negotiated` mean.

Roasbeef committed:
Merge pull request #1086 from ariard/2023-06-specify-max-cltv-expiry-delta

BOLT4: Specify max HTLC nLocktime for expiry_too_far

niftynei committed:
spelling: fix spelling mistakes + add new words to ignore spelling

niftynei committed:
interactive-tx: highlight MUST of serial_id sorting

Suggested-By: @dunxen

niftynei committed:
tx-abort: port over meaning of the `data` field from `warning`

`tx_abort`'s structure comes from the `warning`/`error` messages,
but we failed to port over the rationale/rules for the `data` field.

Suggested-By: @morehouse

niftynei committed:
tx-complete: clarify that the total input/total output incl funding

Found-by: @morehouse

niftynei committed:
nit: verbiage + wording

Suggested-By: @morehouse

niftynei committed:
tx-signatures: address comments regarding underpyament of fees

Reported-By: @morehouse

niftynei committed:
dual-fund, nit: change wording on common input heuristic

Suggested-By: @devrandom

niftynei committed:
dual-fund, wording: move sighash_all into existing verbiage

Suggested-By: @morehouse

niftynei committed:
dual-fund, verbiage: clarify situation for closing a underfunded channel

Moved up some rationale from the Rationale section and added a
bit of clarification to when you'd want to close/cancel an open.

Reported-By: @morehouse

niftynei committed:
dual-fund: add `require_confirmed_inputs` to RBF messages

Make `require_confirmed_inputs` explicit for RBF regnegotiation.

Requested-By: @t-bast

t-bast committed:
Add `min_final_expiry_delta` to blinded route example (#1131)

Payment recipients must ensure that they have a few blocks
before fulfilling the payment, in case the inbound channel
force-closes.

xiaolou86 committed:
Fix typos (#1130)

* BOLT 04: fix typos
* proposals: fix typos

jrakibi committed:
Update Signet port to correct hex value in BOLT #1 spec

rustyrussell committed:
BOLT 4: It's Check Lock Time Verify not Check Time Lock Verify. (#1141)

I always get this wrong too, so CLN actually has a source check for this, and it triggered when importing the latest spec!

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

trigger-xyz committed:
BOLT 7: correct the default `min_final_cltv_expiry_delta` in routing example (#1143)

carlaKC committed:
04-onion-routing: strict validation of scid for blinded payments (#1147)

This commit updates bolt04 to more strictly enforce that encrypted_data
that is part of a blinded payment only has short_channel_id set. On
the reader side, we disallow setting of both short_channel_id and
next_node_id (which is intended for use in the context of onion
messages), and on the writer side we specify that next_node_id should
not be included by recipients.

ErikDeSmedt committed:
Fix broken link in BOLT-2 (#1148)

rustyrussell committed:
BOLT 9: Add `ASSUMED` marker we can use on features.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 9: `option_data_loss_protect` and `option_static_remotekey` are now assumed.

These still have names and numbers, since they appear in `channel_type`.  They are somewhat tangled with each other, so let's tie them together as assumed.

option_data_loss_protect is advertized by all by 11 nodes(*), and option_static_remotekey all but 16 nodes.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
[* there are 449 three-year old LND nodes which advertize `2200` as their features, which have already been trimmed from most gossip for not having htlc_maximum_msat in their channel_updates]

rustyrussell committed:
BOLT 9: Remove option_anchor_outputs, in favor of option_anchors_zero_fee_htlc_tx.

It's supported only by pre-23.05 core-lightning nodes built with
EXPERIMENTAL_FEATURES.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 9: rename option_anchors_zero_fee_htlc_tx to option_anchors.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

Roasbeef committed:
BOLT-04: use underscores in place of parens for blinded paths notation

In this commit, we propose a purely syntactical change to the current
blinded paths specification. Rather than denote the public key of the i-th
node as `E(i)`, we propose that instead it's denoted as: `E_i`. This results
in less overall characters, and is more similar to notation customarily
used in LaTeX.

My personal preference is that the proposed notation is easier to scan at a
glance, and also less ambiguous (doesn't look like a function call).

Roasbeef committed:
BOLT-04: convert blinded paths syntax to use LaTeX rendering

Github now supports inline LaTeX rendering:
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions

In this commit, we use this new feature to improve the rendering of the
routines for blinded paths.

rustyrussell committed:
BOLT 9: assume var_onion_optin.

Advertized as supported by all but 6 nodes (and those can no longer
route payments since people only send the modern onion these days)

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 9: Remove initial_routing_sync.

This only had an effect when `gossip_queries` was not negotiated, which is now assumed.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 9: Assume gossip_queries.

Supported by all but 11 nodes*.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
[* there are 449 three-year old LND nodes which advertize `2200` as their features, which have already been trimmed from most gossip for not having htlc_maximum_msat in their channel_updates]

Roasbeef committed:
Merge pull request #1151 from Roasbeef/blinded-paths-notation

BOLT-04: use underscores in place of parens for blinded paths notation

t-bast committed:
gossip: 12-blocks delay channel closed follow-up (#1156)

We introduced a 12-blocks delay for channel closing in #1004, but we
didn't update the requirement in the section about pruning.

ProofOfKeags committed:
require quiescence sessions to be single-use

t-bast committed:
Clean-up: follow-up on removing spec features

This is a follow-up to https://github.com/lightning/bolts/pull/1092
that fixes the following issues:

- fix a few typos
- remove non-zero-fee anchors test cases
- remove `remote_pubkey` rotation

Roasbeef committed:
Merge pull request #1165 from t-bast/remove-old-features-typos

Clean-up: follow-up on removing spec features

rustyrussell committed:
BOLT 2: quiescence protocol.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>




Header from folded patch 'bolt_2__set_an_initiator_in_quiescence.patch':

BOLT #2: Set an initiator in quiescence.

This is especially useful for protocols such as splicing; for
simplified commitment transactions, there is already an implied
initiator at each point, so having the negotiation at splicing
time would be redundant.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>



Header from folded patch 'option_quiesce__feature_to_support_stfu_method.patch':

option_quiesce: feature to support stfu method.

In practice, sftu is useless unless you have something (e.g. channel_upgrade)
which uses it, but adding a feature is best practice IMHO.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 2: add requirement for disconnect if quiescence takes too long with HTLCs pending.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ziggie1984 committed:
BOLT04: Add rationale for constant error decryption. (#1154)

To avoid timing analysis when decrypting failed payments the sender
should act as if the failure in the route came for the 27th hop.
Also changed the maximum number of hops in the route from 20 (legacy)
to 27 (tlv onion).

rustyrussell committed:
BOLT 4: remove obsolete references to `realm`

This was from the legacy onion, and is no longer present.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: Generalize and create proper requirements for onion decryption

There's currently a *description* of how to decrypt an onion, and some requirements
in forwarding.  But it also applies to onion messages, so:

1. Turn the description into actual enumerated requirements.
2. Ensure the description covers both payload and messaging onions.
3. Include both methods to apply the blinding tweak.
4. Leave the actual handling of the extracted payload (payment vs messaging onion) to those specific sections (e.g. reporting failure)

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 2, BOLT 4: refer to the onion decryption section in update_add_htlc/onion message requirements.

This ties it together, saying what to use as associated data, blinding, and what to do on failure.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 2, 4: rename `blinding` to `path_key`.

Sure, it's used to derive a secret for blinding, but it's also used to derive the key
for encrypted_recipient_data.  It's not used as a blinding factor *directly*.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

TheBlueMatt committed:
Drop the required `channel_update` in failure onions (#1173)

As noted previously, `channel_update`s in the onion failure packets
are massive gaping fingerprintign vulnerabilities - if a node
applies them in a publicly-visible way the err'ing node can easily
identify the sender of an HTLC.

While the updates are still arguably marginally useful for nodes to
use in their pathfinding local to retires of the same payment, this
too will eventually become an issue with PTLCs. Further, we
shouldn't be letting nodes get away with delaying payments by
failing to announce the latest channel parameters or enforcing new
parameters too soon, so treating the node as having indicated
insufficient liquidity (or other general failure) is appropriate
in the general case.

Thus, here, we begin phasing out the `channel_update` field,
requiring nodes ignore it outside of the current payment and making
it formally optional (though nodes have been doing this for some
time due to various bugs).

Because some nodes may want to use update data on mobile when they
have stale gossip data, it is left optional.

ellemouton committed:
bolt4+proposals: fix `max_cltv_expiry` calculation (#1176)

* BOLT4: include `min_final_cltv_expiry_delta` in `max_cltv_expiry` calc

Include `min_final_cltv_expiry_delta` in the `max_cltv_expiry`
calculation. Also add a note that indicates that this field may be set
for the final node too. This is useful for the final node as then it
does not need to persist the path expiry separately and can rely on just
checking the `payment_relay` field when the payment arrives.

* BOLT4: include calculation for `total_cltv_delta` of a blinded path

Include an explicit formula to use for determining the total CLTV delta
of a blinded path so that it is clear that it should include the
recipient's `min_final_cltv_expiry_delta`.

* proposals: fix `max_cltv_expiry` value for final hop in example

More info
[here](https://github.com/lightning/bolts/issues/1174#issue-2371364610)
outlining why the example needed to be updated.

t-bast committed:
Clean-up nits

This commit doesn't change the logic at all, it simply:

- removes `realm` from onion test vector
- cleans-up markdown formatting and indents
- fixes typos and missing parenthesis
- consistently uses `_` instead of `-` for field names
- fixes math formatting (including changes from #1169 and #1158)

rustyrussell committed:
BOLT 4: move Route Blinding section up a level, and put `blinded_path` type there.

It's currently buried in the onion message section, but it applies to payments too.

We now have a separate sub-section for the encrypted_data_tlv definition.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: concretely refer to the `blinded_path` type and field when constructing.

This spec was initially written before the `blinded_path` type
existed.  Be precise (and we no longer need to say "MUST communicate"!).

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: clarify blinded path requirements.

Writer parts:
1. Be explicit that the writer creates a route.
2. Make it clear we create shared secrets, then derive blinded points.
3. Refer explicitly to all `blinded_path` fields.

Split reader into the *two* readers:
1. The reader of the blinded path, who uses it to make an onion (which wasn't described at all!)
2. The reader of the encrypted_recipient_data, who decrypts it.

In the latter case, we don't have to discuss unblinding the onion since
that's now covered in the "Onion Decryption" section.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: try to improve blinded path description.

It's a bit complex, but try to convey the idea of an introduction point,
blinded node ids and encrypted blobs.  Since the requirements detail the
two ways to reach the introduction node, I handwaved on that a bit.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: put dummy hop recommendation into the requirements.

It was mentioned in the rationale only.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: rename `onionmsg_hop` to `blinded_path_hop`

It's used for blinded payment paths, too, not just onion messages.

Suggested-by: @t-bast
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: More clarifying changes.

1. Always use the term `encrypted_recipient_data` for the encrypted field:
   it's confusing enough without multiple names!
2. Don't give an option for joining blinded payments, since everyone will
   use an unblinded payment to the introduction node (at least, for now).
3. Avoid the term "payer" and at least note that encrypted_recipient_data
   can be made by the sender themselves, pointing out that the forwarding
   node cannot tell.

Thanks t-bast and gijswijs for this feedback!

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 4: fix nomenclature in bolt04/route-blinding-test.json

Bastien points out we didn't update this when we changed terms.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

alexbosworth committed:
Make blinded link work (#1190)

Fix link reference to point at sub section with minimal link reference

Roasbeef committed:
Merge pull request #1182 from rustyrussell/clarify-onions-part-2

Clarify onions part 2: a bit deeper rework

rustyrussell committed:
BOLT 3: Fix example keys in test vectors.

All the test vectors use static keys now, which are listed above
already as the local_payment_basepoint and remote_payment_basepoint.
The keys listed here are the pre-static rotated ones: if you use
these, the vectors don't work!

We actually use the same basepoint for the HTLCs, but never spelled it
out.  So do that now, and these are the local/remote htlc keys.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

yyforyongyu committed:
trivial: fix wording and links in bolt03 (#1192)

* bolt03: fix link to `bitcoind`'s `policy.cpp` file

The original link is broken, we now replace it with a permalink.

* bolt03: remove 0 `txout` count in closing tx

Txns cannot have zero outputs.

* bolt03: remove wrong ref to `remotepubkey` and fix format

`remotepubkey` has been placed at the wrong place, also fixes the format
so it's easier to read.

morehouse committed:
Replace "logical OR" with "bitwise OR" (#991)

This makes more sense given the context of combining bitmaps.

hugomd committed:
Bolt 11: add example SHA256 hash of description (#994)

Clarify test vector.
